GOVERNANCE_DOC.md

```diff
@@ -0,0 +1,597 @@
+
+
+## Implementation
+
+*Disclaimer: This is a suggestion. Only structs and pseudocode. Actual logic and implementation might widely differ*
+
+### Procedures
+
+`Procedures` define the rule according to which votes are run. There can only be one active procedure at any given time. If governance wants to change a procedure, either to modify a value or add/remove a parameter, a new procedure has to be created and the previous one rendered inactive.
+
+```Go
+type Procedure struct {
+  VotingPeriod      uint64              //  Length of the voting period. Initial value: 2 weeks
+  MinDeposit        uint64              //  Minimum deposit for a proposal to enter voting period. 
+  OptionSet         []string            //  Options available to voters. {Yes, No, NoWithVeto, Abstain}
+  ProposalTypes     []string            //  Types available to submitters. {PlainTextProposal, SoftwareUpgradeProposal}
+  Threshold         uint64              //  Minimum value of Yes votes to No votes ratio for proposal to pass. Initial value: 0.5
+  Veto              rational.Rational   //  Minimum value of Veto votes to Total votes ratio for proposal to be vetoed. Initial value: 1/3
+  MaxDepositPeriod  uint64              //  Maximum period for Atom holders to deposit on a proposal. Initial value: 2 months
+  GovernancePenalty uint64              //  Penalty if validator does not vote
+  
+  ProcedureNumber   uint16              //  Incremented each time a new procedure is created
+  IsActive          bool                //  If true, procedure is active. Only one procedure can have isActive true.
+}
+```
+
+### Proposals
+
+`Proposals` are item to be voted on. They can be submitted by any Atom holder via a `TxSubmitProposal` transaction.
+
+```Go
+type TxSubmitProposal struct {
+  Title           string        //  Title of the proposal
+  Description     string        //  Description of the proposal
+  Type            string        //  Type of proposal. Initial set {PlainTextProposal, SoftwareUpgradeProposal}
+  Category        bool          //  false=regular, true=urgent
+  InitialDeposit  uint64        //  Initial deposit paid by sender. Must be strictly positive.
+}
+
+type Proposal struct {
+  Title             string              //  Title of the proposal
+  Description       string              //  Description of the proposal
+  Type              string              //  Type of proposal. Initial set {PlainTextProposal, SoftwareUpgradeProposal}
+  Category          bool                //  false=regular, true=urgent
+  Deposit           uint64              //  Current deposit on this proposal. Initial value is set at InitialDeposit
+  SubmitBlock       uint64              //  Height of the block where TxSubmitProposal was included
+  VotingStartBlock  uint64              //  Height of the block where MinDeposit was reached. -1 if MinDeposit is not reached.
+  Votes             map[string]uint64   //  Votes for each option (Yes, No, NoWithVeto, Abstain)
```

Comment:
don't use a map, the type should probably be either an array of int64 of a new struct `VoteRes`

```golang
type VoteRes struct {
    Yes, No, NoWithVeto, Abstain int64
}
```

Reply:
(can't use maps in a merkle store due to non-determinism)

Reply:
@sunnya97 what's your opinion? Array of new struct?

Reply:
What are we mapping from? So what would the `string` be? Validator/Delegator keys?

Also, does the proposal struct handle the overriding by delegators and can people update their votes?

Reply:
- string would be each option `Yes` `No` `NoWithVeto` `Abstain`.
- overriding is done when processing `TxVote`
- In MVP there is no vote update

Reply:
With this structure we can't attribute who submitted what vote though? How are we going to update the votes when processing TxVote?

Reply:
There are 4 additional lists linked to each `proposalID`. One of them is `VotersList` where you can find every pubkey that voted on which option and associated to which validator. 

Originally I wanted a mapping PubKey -> {option, valPubKey} instead of `voterList` but apparently not possible in Go. So I put lists as general struct but surely this needs to be adapted as @rigelrozanski said. Probably @sunnya97 should have a say in this as he is going to implement it.

Reply:
@adrianbrink @gamarin2 We don't need to be able to dicern who voted for what... we just need to know that nobody is voting twice and that the sum of the votes made... right

Reply:
We need to be able to dicern:
- Who voted under which validator (the same PubKey can vote twice if it has bonded Atoms to two distincts validators)
- What option did each validator vote for